File: echolot-demo/src/main/java/de/exxcellent/echolot/testapp/testscreen/PieChartTest.java
/*
 * This file (PieChartTest.java) is part of the Echolot Project (hereinafter "Echolot").
 * Copyright (C) 2008-2010 eXXcellent Solutions GmbH.
 *
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 */

package de.exxcellent.echolot.testapp.testscreen;

import de.exxcellent.echolot.app.PieChart;
import de.exxcellent.echolot.event.PieSectorSelectEvent;
import de.exxcellent.echolot.listener.PieSectorSelectListener;
import de.exxcellent.echolot.model.AnimationType;
import de.exxcellent.echolot.model.Location;
import de.exxcellent.echolot.model.PieModel;
import de.exxcellent.echolot.model.PieSector;
import de.exxcellent.echolot.testapp.ButtonColumn;
import de.exxcellent.echolot.testapp.StyleUtil;
import nextapp.echo.app.*;
import nextapp.echo.app.event.ActionEvent;
import nextapp.echo.app.event.ActionListener;

import javax.swing.*;

/**
 * This class represents a visual Test-Area for the PieChart and all it's properties
 *
 * @author Ralf Enderle
 */
public class PieChartTest extends SplitPane {

    final PieChart pieChart;

    /**
     * Visual Test for the Pie-Chart
     */
    public PieChartTest() {
        super(SplitPane.ORIENTATION_HORIZONTAL, new Extent(250, Extent.PX));
        setStyleName("DefaultResizable");

        ButtonColumn controlsColumn = new ButtonColumn();
        controlsColumn.setStyleName("TestControlsColumn");
        add(controlsColumn);

        pieChart = new PieChart();
        pieChart.setWidth(300);
        pieChart.setHeight(200);
        pieChart.setShowLegend(true);
        Font font = new Font(Font.SERIF, Font.ITALIC, new Extent(10));
        pieChart.setFont(font);
        pieChart.setForeground(Color.ORANGE);
        pieChart.setShowLegend(true);
        pieChart.setShowPopUp(true);
        pieChart.setDoAnimation(true);
        pieChart.setSectorAbbrevShow(false);
        pieChart.setDoClientSorting(false);

        pieChart.setFallbackSectorColor0(Color.DARKGRAY);
        pieChart.setFallbackSectorColor1(Color.ORANGE);
        pieChart.setFallbackSectorColor2(Color.MAGENTA);

        pieChart.setPieModel(getPieModelDefault());

        pieChart.addPieSectorSelectListener(new PieSectorSelectListener() {
            public void sectorSelection(PieSectorSelectEvent e) {
                System.out.println("You selected: SectorName:'" + e.getPieSector().getName() + "' SectorValue:'" + e.getPieSector().getValue() +
                        "' Defined UserObject:'" + e.getPieSector().getUserObject() + "'");
            }
        });

        add(pieChart);

        // -- sectors ---
        controlsColumn.add(new Label("Different number of sectors"));
        controlsColumn.addButton("Default PieModel", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setPieModel(getPieModelDefault());
            }
        });

        controlsColumn.addButton("PieModel with more sectors", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setPieModel(getPieModel_MoreSectors());
            }
        });

        controlsColumn.addButton("set PieModel to 'null'", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setPieModel(null);
            }
        });

        // --- color ---
        controlsColumn.add(new Label("sector color"));
        controlsColumn.addButton("Random Background-Color ", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setBackground(StyleUtil.randomColor());
            }
        });

        // --- width/height ---
        controlsColumn.add(new Label("Width/Height"));
        controlsColumn.addButton("Width: 200 ", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setWidth(200);
            }
        });
        controlsColumn.addButton("Width: 300", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setWidth(300);
            }
        });

        controlsColumn.addButton("Width: 400", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setWidth(400);
            }
        });
        controlsColumn.addButton("Height: 200", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setHeight(200);
            }
        });
        controlsColumn.addButton("Height: 300", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setHeight(300);
            }
        });
        controlsColumn.addButton("Height: 400", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setHeight(400);
            }
        });

        // Popup-Styling
        controlsColumn.add(new Label("Popup-Styling"));
        controlsColumn.addButton("Random border-color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomBorderColor();
            }
        });

        controlsColumn.addButton("Random background-color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomBackgroundColor();
            }
        });

        controlsColumn.addButton("Random foreground-color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomForegroundColor();
            }
        });

        controlsColumn.addButton("Random font", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomFont();
            }
        });

        // --- Legend ---
        controlsColumn.add(new Label("Legend-Position"));
        controlsColumn.addButton("Legend: EAST", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendPosition(Location.EAST);
            }
        });
        controlsColumn.addButton("Legend: SOUTH", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendPosition(Location.SOUTH);
            }
        });
        controlsColumn.addButton("Legend: NORTH - Not yet supported!", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendPosition(Location.NORTH);
            }
        });
        controlsColumn.addButton("Legend: WEST - Not yet supported!", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendPosition(Location.WEST);
            }
        });

        controlsColumn.addButton("Random legend-Color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setForeground(StyleUtil.randomColor());
            }
        });
        controlsColumn.addButton("Random legend-Font", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setFont(StyleUtil.randomFont());
            }
        });
        controlsColumn.addButton("LegendGap:default '1.2'", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendGapFactor(1.2);
            }
        });
        controlsColumn.addButton("LegendGap: '0.8'", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendGapFactor(0.8);
            }
        });
        controlsColumn.addButton("LegendGap:default '2'", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setLegendGapFactor(2);
            }
        });
        // -- Abbreviation ---
        controlsColumn.add(new Label("Abbreviation"));
        controlsColumn.addButton("Toggle Abbreviation ON/OFF", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setSectorAbbrevShow(!pieChart.isSectorAbbrevShow());
            }
        });
        // -- animation ---
        controlsColumn.add(new Label("Animation"));
        controlsColumn.addButton("Toggle animation ON/OFF", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setDoAnimation(!pieChart.isDoAnimation());
            }
        });
        controlsColumn.addButton("AnimationType: ELASTIC", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setAnimationType(AnimationType.ELASTIC);
            }
        });
        controlsColumn.addButton("AnimationType: BOUNCE", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setAnimationType(AnimationType.BOUNCE);
            }
        });
        controlsColumn.addButton("AnimationType: BACK_OUT", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setAnimationType(AnimationType.BACK_OUT);
            }
        });

        // -- sorting --
        controlsColumn.add(new Label("Client sorting"));
        controlsColumn.addButton("Toggle clientSorting", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                pieChart.setDoClientSorting(!pieChart.isDoClientSorting());
            }
        });


    }

    private PieModel getPieModelDefault() {
        PieSector sector_1 = new PieSector("Foo", 50, "value: 50", Color.GREEN, "My UserObject for Foo");
        PieSector sector_2 = new PieSector("Bar", 80, "Hi there \n The value is 80", Color.RED, "My UserObject for Bar");
        return new PieModel(new PieSector[]{sector_1, sector_2});
    }

    private PieModel getPieModel_MoreSectors() {
        PieSector sector_1 = new PieSector("Test", 70, "Test", "T", null, Color.LIGHTGRAY, null);
        PieSector sector_2 = new PieSector("Blubber", 20, "Blubber", "Bl", null, null);
        PieSector sector_3 = new PieSector("Bar", 150, "Bar", "Ba", null, null);
        PieSector sector_4 = new PieSector("Foo", 0, "Bar", "Ba", null, Color.WHITE, null);
        return new PieModel(new PieSector[]{sector_1, sector_2, sector_3, sector_4});
    }

    private void setRandomBorderColor() {
        pieChart.setPopupBorderColor(StyleUtil.randomBrightColor());
    }

    private void setRandomBackgroundColor() {
        pieChart.setPopupBackground(StyleUtil.randomBrightColor());
    }

    private void setRandomForegroundColor() {
        pieChart.setPopupForeground(StyleUtil.randomBrightColor());
    }

    private void setRandomFont() {
        pieChart.setPopupFont(StyleUtil.randomFont());
    }

}

File: echolot-webcontainer/src/test/web/testscreen/PieChartTest.js
/*
 * This file (PieChartTest.js) is part of the Echolot Project (hereinafter "Echolot").
 * Copyright (C) 2008-2011 eXXcellent Solutions GmbH.
 *
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 */

/**
 * Testclass for the PieChart
 *
 * @author Ralf Enderle
 */
exxcellent.test.PieChartTest = Core.extend({

    _pieChart: null,
    _mainContent: null,

    $construct: function (testArea, mainContent) {
        this._mainContent = mainContent;

        var controlColumn = new Echo.Column({
            styleName: "TestControl",
            children: [
                new Echo.Label({
                    text: "PieChart", styleName: "Title"
                }),
                new Echo.Label({
                    text: "Models", styleName: "Default"
                }),
                this._createButton("Model 1 - 3 sectors",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.PIE_MODEL, this._createPieModel_1().pieModel);
                    }
                ),
                this._createButton("Model 2 - 3 sectors",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.PIE_MODEL, this._createPieModel_2().pieModel);
                    }
                ),
                this._createButton("Model 3 - More Sectors",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.PIE_MODEL, this._createPieModel_3().pieModel);
                    }
                ),
                this._createButton("Model 4 - 3 Sectors",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.PIE_MODEL, this._createPieModel_4().pieModel);
                    }
                ),
                new Echo.Label({
                    text: "Size", styleName: "Default"
                }),
                this._createButton("Height +10",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.HEIGHT, this._pieChart.get(exxcellent.PieChart.HEIGHT) + 10);
                    }
                ),
                this._createButton("Height -10",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.HEIGHT, this._pieChart.get(exxcellent.PieChart.HEIGHT) - 10);
                    }
                ),
                new Echo.Label({
                    text: "Animation", styleName: "Default"
                }),
                this._createButton("Toggle Animation",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.DO_ANIMATION, !this._pieChart.get(exxcellent.PieChart.DO_ANIMATION));
                    }
                ),
                this._createButton("Bounce",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.ANIMATION_TYPE, "bounce");
                    }
                ),
                new Echo.Label({
                    text: "Styling", styleName: "Default"
                }),
                this._createButton("toggle sectorAbbrev ON/OFF ",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.SECTOR_ABBREV_SHOW, !this._pieChart.get(exxcellent.PieChart.SECTOR_ABBREV_SHOW));
                    }
                ),
                this._createButton("toggle clientSorting ON/OFF ",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.DO_CLIENT_SORTING, !this._pieChart.get(exxcellent.PieChart.DO_CLIENT_SORTING));
                    }
                ),
                this._createButton("toggle hide slices with value 0 from legend ON/OFF ",
                    function () {
                        this._pieChart.set(exxcellent.PieChart.LEGEND_HIDE_ZERO_VALUES, !this._pieChart.get(exxcellent.PieChart.LEGEND_HIDE_ZERO_VALUES));
                    }
                )
            ]
        });

        this._pieChart = this._createPieChart();
        this._pieChart.set(exxcellent.PieChart.PIE_MODEL, this._createPieModel_1().pieModel);

        var splitPane = new Echo.SplitPane({
            styleName: "TestControl",
            children: [
                controlColumn,
                new Echo.Column({
                    styleName: "Default",
                    children: [this._pieChart]
                })
            ]
        });
        testArea.add(splitPane);
    },

    _createPieChart: function () {
        return new exxcellent.PieChart({
            font: {
                size: "18",
                typeface: ["Courier New", "Courier", "Monospace"],
                italic: true
            },
            fallbackSectorColor_0: 'red',
            fallbackSectorColor_1: 'green',
            fallbackSectorColor_2: 'blue',
            fallbackSectorColor_3: 'maroon',
            foreground: '#aaffaa',
            legendPosition: 'east',
            showLegend: true,
            showPopUp: true,
            doAnimation: true,
            width: 400,
            height: 200,
            events: {
                pieSectorSelect: Core.method(this, this._dummyAction) }
        });
    },

    /**
     * Creates a Pie model with a few sectors
     */
    _createPieModel_1: function () {
        // the sectors
        var sectors = [
            new exxcellent.model.PieSector("Blubber", 260, "Blubber PopUp", true, '#33FF99', 'Abbrev 1', "#000", 10)
            ,
            new exxcellent.model.PieSector("Bar", 1000, "Bar PopUp", true, '#666633', 'Abbrev 3', null, "#000", 12)
            ,
            new exxcellent.model.PieSector("One", 1, "one", true, '#666633', 'Abbrev 3', null, "#000", 12)
            ,
            new exxcellent.model.PieSector("Bar", 500, "Bar PopUp", true, '#666633', 'Abbrev 3', null, "#000", 12)
            ,
            new exxcellent.model.PieSector("Also one", 1, "also one", true, '#666633', 'Abbrev 3', null, "#000", 12)
            ,
            new exxcellent.model.PieSector("zero2", 0, "zero", true, '#666633', 'Abbrev 3', null, "#000", 12)
        ];

        // return the whole thing
        return {
            pieModel: new exxcellent.model.PieModel(sectors)
        };
    },
    /**
     * Creates a Pie model with a few sectors
     */
    _createPieModel_2: function () {
        // the sectors
        var sectors = [
            new exxcellent.model.PieSector("A", 90, "A PopUp", true, '#33FF99'),
            new exxcellent.model.PieSector("B", 60, "B PupUp", true, '#33AA99'),
            new exxcellent.model.PieSector("C", 90, "C PopUp", true, '#33BB99')
        ];
        var pieModel = new exxcellent.model.PieModel(sectors);
        // return the whole thing
        return {
            pieModel: pieModel

        };
    },
    /**
     * Creates a Pie model with a few sectors
     */
    _createPieModel_3: function () {
        // the sectors
        var sectors = [
            new exxcellent.model.PieSector("Ba", 90, "PopUp 0", true),
            new exxcellent.model.PieSector("Be", 60, "PopUp 1", true),
            new exxcellent.model.PieSector("Bi", 120, "PopUp 2", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 3", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 4", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 5", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 6", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 7", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 8", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 9", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 10", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 11", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 12", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 13", true),
            new exxcellent.model.PieSector("Bo", 40, "PopUp 14", true),
            new exxcellent.model.PieSector("Bu", 20, "PopUp 15", true)

        ];
        var pieModel = new exxcellent.model.PieModel(sectors);
        pieModel.radius = 150;
        // return the whole thing
        return {
            pieModel: pieModel
        };
    },

    /**
     * Creates a Pie model with a few sectors
     */
    _createPieModel_4: function () {
        // the sectors
        var sectors = [
            new exxcellent.model.PieSector("Blubber", 80, "PopUp", true, '#11DD99'),
            new exxcellent.model.PieSector("Foo", 100, "PopUp", true, '#33AA66'),
            new exxcellent.model.PieSector("Bar", 300, "PopUp", true, '#33CC22')
        ];
        var pieModel = new exxcellent.model.PieModel(sectors);
        pieModel.legendPos = 'south';
        // return the whole thing
        return {
            pieModel: pieModel
        };
    },

    /**
     * Helper
     * @param text
     * @param action
     */
    _createButton: function (text, action) {
        return new Echo.Button({
            text: text,
            styleName: "Default",
            events: {
                action: Core.method(this, action)
            }
        });
    },
    _dummyAction: function (event) {
        if (event && window.console) {
            console.log("Action triggered event: " + event.type + " data: " + event.data);
        }
        this._mainContent.showMsg("Action triggered!", event.type + " data: " + event.data);
        return false;
    }

});